### PR TITLE
SFTTrainer: merge entropy and accuracy computation to eliminate redundant logits copy

### PR DESCRIPTION
**Pull request by flutist**
**PR Title**

`SFTTrainer: merge entropy and accuracy computation to eliminate redundant logits copy`

**PR Description**

## What does this PR do?

Merges the entropy and token accuracy computations in `SFTTrainer.compute_loss` into a single `torch.no_grad()` block, eliminating a redundant `.contiguous()` copy of the full logits tensor.

When `loss_type="nll"` (the default) and Liger kernel is not used, `compute_loss` previously materialized `shift_logits = outputs.logits[..., :-1, :].contiguous()` **twice** — once for entropy calculation and once for accuracy calculation. The logits tensor shape is `[batch_size, seq_len, vocab_size]` — for a model with `vocab_size` ~152k (e.g. Qwen2.5), this is one of the largest allocations in `compute_loss`. The second copy is entirely redundant since both metrics operate on the same shifted logits.

### Changes:

- Moved `predictions = shift_logits.argmax(dim=-1)` and accuracy computation into the existing entropy `torch.no_grad()` block, reusing the same `shift_logits` tensor
- Removed the separate accuracy `else` branch that performed a second `.contiguous()` call
- Net reduction of ~20 lines of code

## Memory benchmark

Benchmark script (not included in this PR):

```python
import gc
import torch
from datasets import load_dataset
from trl import SFTConfig, SFTTrainer

if not torch.cuda.is_available():
    print("ERROR: No CUDA GPU available.")
    exit(1)

gc.collect()
torch.cuda.empty_cache()
torch.cuda.reset_peak_memory_stats()

dataset = load_dataset("trl-lib/Capybara", split="train[:200]")

args = SFTConfig(
    output_dir="/tmp/sft_peak_bench",
    per_device_train_batch_size=4,
    max_steps=3,
    max_length=1024,
    gradient_checkpointing=True,
    bf16=True,
    report_to="none",
    logging_steps=1,
    loss_type="nll",
)

trainer = SFTTrainer(model="Qwen/Qwen2.5-0.5B-Instruct", args=args, train_dataset=dataset)
trainer.train()

peak_mb = torch.cuda.max_memory_allocated() / 1024**2
print(f"\nPeak allocated: {peak_mb:.1f} MB")
```

Results (Qwen2.5-0.5B-Instruct, batch\_size=4, max\_length=1024, bf16, gradient\_checkpointing=True):

| | Peak allocated |
|---|---|
| Before | 15799 MB |
| After | 13650 MB |
| **Saved** | **2149 MB (−13.6%)** |

No functional regression — loss, entropy, and accuracy metrics are consistent across both runs:

| Step | Loss (before) | Loss (after) | Accuracy (before) | Accuracy (after) |
|------|--------------|-------------|-------------------|-----------------|
| 1 | 2.481 | 2.481 | 0.5254 | 0.5254 |
| 2 | 2.149 | 2.149 | 0.5424 | 0.5424 |
| 3 | 1.415 | 1.416 | 0.6437 | 0.6441 |

(Minor differences in step 3 are expected due to non-deterministic operations; steps 1–2 are identical, confirming no computational change.)

---

> [!NOTE]
> **Low Risk**
> Training-time metric logging only; loss path and Liger/chunked_nll behavior are unchanged aside from deduplicating logits handling for NLL.
> 
> **Overview**
> For default NLL training **without** the Liger kernel, **mean token accuracy** is now computed in the same `torch.no_grad()` block as **entropy**, reusing one shifted, `.contiguous()` logits tensor instead of building it twice.
> 
> The standalone accuracy branch (duplicate shift/mask/argmax/gather logic) is removed. **Liger** still logs accuracy from `outputs.token_accuracy` when present; **chunked_nll** is unchanged.
> 
> This is a memory/throughput optimization for large vocabularies—reported peak CUDA allocation drops on the order of ~10–15% with matching loss/accuracy on benchmark steps.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 89f11ac89de52d07d360389f4a6a1b98eef1d464. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by flutist** [comment]
@qgallouedec Hi! Could you take a look at this PR when you have a chance?

File: trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -1631,16 +1631,26 @@ def compute_loss(self, model, inputs, return_outputs=False, num_items_in_batch=N
                 ):
                     shift_logits = shift_logits[:, self.num_virtual_tokens :, :]
 
+                # Compute entropy and accuracy together to avoid a second .contiguous() copy of logits
```
Comment (qgallouedec) [inline trl/trainer/sft_trainer.py]:
```suggestion
```

it makes sense in the context of this pr, but not standalone imo

**Review by qgallouedec — APPROVED** [review]
LGTM! big fan of this kind of simplifications!

**Comment by flutist** [comment]
Thanks a lot, and could you help to approve this pr too?@qgallouedec https://github.com/huggingface/trl/pull/5882
